### PR TITLE
8286302: Port JEP 425 to PPC64

### PR DESCRIPTION
**Pull request by reinrich**
Hi,

this is the port of [JEP 425: Virtual Threads (Preview)](https://openjdk.org/jeps/425)) to PPC64.
More precisely it is the port of vm continuations in hotspot to PPC64. It allows to run with `-XX:+VMContinuations` which is a prerequisit for 'real' virtual threads (oxymoron?).

Most of the shared code changes are related to a new platform dependent constant `frame::metadata_words_at_top`. It is either added or subtracted to a frame address or size.

The following is supposed to explain (without real life details) why it is needed in addition to the existing `frame::metadata_words`. The diagram shows a frame at `SP` and its stack arguments. The caller frame is located at `callers_SP`.

```
  X86 / AARCH64                             PPC64:

  :                 :                       :                 :
  :                 :                       :                 :
  |                 |                       |                 |
  |-----------------|                       |-----------------|
  |                 |                       |                 |
  | stack arguments |                       | stack arguments |
  |                 |<- callers_SP          |                 |
  ===================                       |-----------------|
  |                 |                       |                 |
  | metadata@bottom |                       | metadata@top    |
  |                 |                       |                 |<- callers_SP
  |-----------------|                       ===================
  |                 |                       |                 |
  |                 |                       |                 |
  |                 |                       |                 |
  |                 |                       |                 |
  |                 |<- SP                  |                 |
  ===================                       |-----------------|
                                            |                 |
                                            | metadata@top    |
                                            |                 |<- SP
                                            ===================
```

On X86 and AARCH64 metadata (that's return address, saved SP/FP, etc.) is stored at the frame bottom (`metadata@bottom`). On PPC64 it is stored at the frame top (`metadata@top`) where it affects size and address calculations. Shared code deals with this by making use of the platform dependent constant `frame::metadata_words_at_top`.

* size required to 'freeze' a single frame with its stack arguments in a `StackChunk`:
  `sizeof(frame) + sizeof(stack arguments) + frame::metadata_words_at_top`

* address of stack arguments:
  `callers_SP + frame::metadata_words_at_top`

* The value of `frame::metadata_words_at_top` is 0 words on X86 and AARCH64. On PPC64 it is 4 words.

Please refer to comments I've added for more details (e.g. the comment on StackChunkFrameStream<frame_kind>::frame_size()). Recently I've given a talk about vm continuations and the PPC64 port to my colleagues. It's rather an overview than a deep dive. [The slides](http://cr.openjdk.java.net/~rrich/webrevs/2022/8286302/202210_loom_ppc64_port.pdf) might serve as well as an intro to the matter.

The pr includes the new test jdk/jdk/internal/vm/Continuation/BasicExp.java which I wrote while doing the port. The test cases vary from simple to not-so-simple. One of the main features is that a CompilationPolicy passed as argument controls which frames are compiled/interpreted when freezing by defining a sliding window of compiled / interpreted frames which produces interesting transitions with and without stack arguments. There is overlap with Basic.java and Fuzz.java. Let me know wether to remove or keep BasicExp.java. Runtime with fastdebug: 2m on Apple M1 Pro and 3m on Intel Xeon E5-2660 v3 @ 2.60GHz. Note that -XX:+VerifyContinuations is explicitly set as a found it very useful, it increases the runtime quite a bit though.

Testing: the change passed our CI testing: most JCK and JTREG tests, also in Xcomp mode, SPECjvm2008, SPECjbb2015, Renaissance Suite and SAP specific tests with fastdebug and release builds on the standard platforms plus PPC64. These tests do include hotspot_loom and jdk_loom JTREG tests which I've also run with TEST_VM_OPTS="-XX:+VerifyContinuations" on X86_64, PPC64, and AARCH64.

Thanks, Richard.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8286302](https://bugs.openjdk.org/browse/JDK-8286302): Port JEP 425 to PPC64

### Reviewers
 * [Tyler Steele](https://openjdk.org/census#tsteele) (@backwaterred - Committer) ⚠️ Review applies to [c1d2f878](https://git.openjdk.org/jdk/pull/10961/files/c1d2f8787d29cecf8e1fad320c4dc48a7c2a424a)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to [22430750](https://git.openjdk.org/jdk/pull/10961/files/224307503d0abbf34720275e851501b47f175a96)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10961/head:pull/10961` \
`$ git checkout pull/10961`

Update a local copy of the PR: \
`$ git checkout pull/10961` \
`$ git pull https://git.openjdk.org/jdk pull/10961/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10961`

View PR using the GUI difftool: \
`$ git pr show -t 10961`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10961.diff">https://git.openjdk.org/jdk/pull/10961.diff</a>

</details>

### CODE REVIEW

**Comment by RealFYang** [comment]
Hi, I also tested this on my Linux RISC-V platform. Test results look good. Thanks. 
- Tier1-4 tests
- Tests under test/jdk/java/lang/Thread/virtual with extra options "-XX:+VerifyContinuations -XX:+VerifyStack"

**Comment by reinrich** [comment]
> Hi, I also tested this on my Linux RISC-V platform which has support for virtual threads. Test results look good. Thanks.
> 
>     * Tier1-4 tests
> 
>     * Tests under test/jdk/java/lang/Thread/virtual with extra options "-XX:+VerifyContinuations -XX:+VerifyStack"
> 
>     * Non-trivial benchmark workloads, like Renaissance, SPECjvm2008, SPECjbb2015, etc.

Great. Thanks for the testing.

Richard.

**Comment by backwaterred** [comment]
I see a couple failures on Linux/ppc64le when testing with `-XX:+VerifyContinuations`.

jdk/internal/vm/Continuation/Fuzz.java#default 
jdk/internal/vm/Continuation/Fuzz.java#preserve-fp 

It may be reasonable to add these to a ProblemList.txt and address them at a different time.

**Comment by reinrich** [comment]
> I see a couple failures on Linux/ppc64le when testing with `-XX:+VerifyContinuations`.
> 
> jdk/internal/vm/Continuation/Fuzz.java#default jdk/internal/vm/Continuation/Fuzz.java#preserve-fp
> 
> It may be reasonable to add these to a ProblemList.txt and address them at a different time.

Thanks for looking at the pr. The failures are due to timeouts. Depending on the load on your test system it might be necessary to increment the timeout factor (JTREG="TIMEOUT_FACTOR=8"). There were issues with the test before (see https://bugs.openjdk.org/browse/JDK-8290211).

**Comment by backwaterred** [comment]
The failures are indeed time-dependant, and pass with JTREG="TIMEOUT_FACTOR=16".

**Review by backwaterred — APPROVED** [review]
These changes are significant. They appear well thought out and well executed. Thanks for submitting this PR.

File: test/jdk/jdk/internal/vm/Continuation/BasicExp.java
Comment (turbanoff) [inline test/jdk/jdk/internal/vm/Continuation/BasicExp.java:483]: nit
```suggestion
                } catch (UnhandledException e) {
```

File: test/jdk/jdk/internal/vm/Continuation/BasicExp.java
Comment (turbanoff) [inline test/jdk/jdk/internal/vm/Continuation/BasicExp.java:341]: nit
```suggestion
            if (compWindowMode == CompWindowMode.NO_COMP_WINDOW) return false;
```

File: test/jdk/jdk/internal/vm/Continuation/BasicExp.java
Comment (turbanoff) [inline test/jdk/jdk/internal/vm/Continuation/BasicExp.java:201]: nit
```suggestion
            compPolicySelection = Integer.parseInt(args[0]);
```

File: src/hotspot/share/oops/stackChunkOop.inline.hpp
Comment (reinrich) [reply src/hotspot/share/oops/stackChunkOop.inline.hpp:126]: > Your change here looks good, but the assertion condition seems incorrect. If `(sp() == stack_size()) == false` and `(sp() >= stack_size() - argsize() - frame::metadata_words_at_top) == false`, then the assertion passes. Unless there is a case for this behaviour, I think it's safe to change this comparison to logical AND.

That's intended.

The assertion checks the equivalence of

`sp() == stack_size()` and `sp() >= stack_size() - argsize() - frame::metadata_words_at_top`

If the first predicate is false (meaning the stack is not empty) then the 2nd predicate must be false also.

Note there is an alternative `is_empty()` method that uses the 2nd predicate: https://github.com/openjdk/jdk/blob/499406c764ba0ce57079b1f612297be5b148e5bb/src/hotspot/share/runtime/continuationFreezeThaw.cpp#L437-L441



Maybe the following diagram of the stack in the StackChunk is useful:


```
                          Offset: stack_size()
      ===================
      |                 |
      | Stack Arguments |
      | to Bottom Frame | Offset: stack_size() - argsize()
      |-----------------|
      |                 |
      | metadata@top    |
      |                 | Offset: stack_size() - argsize() - frame::metadata_words_at_top
      ===================
      | Bottom Frame    |
      |                 |
      |                 |
      ===================
      |                 |
      :                 :
      : FRAMES          :
      :                 :
      |                 |
      ===================
      | Top Frame       |
      |                 |
      |                 |
      |-----------------|
      |                 |
      | metadata@top    |
      |                 | Offset: sp()
      ===================
      :                 :
      :                 :
      : Free Space      :
      :                 :
      :                 :
      :                 : Offset: 0
      ...................
      
Offsets are relative to stackChunkOopDesc::start_of_stack()
```

File: src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp
Comment (reinrich) [reply src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp]: Done

**Comment by reinrich** [comment]
> These changes are significant. They appear well thought out and well executed. Thanks for submitting this PR.

Thanks for your review @backwaterred

File: test/jdk/jdk/internal/vm/Continuation/BasicExp.java
Comment (reinrich) [reply test/jdk/jdk/internal/vm/Continuation/BasicExp.java]: Done. Thanks for looking at the PR!

File: src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp:1803]: Would you like to align the indentation?

File: src/hotspot/cpu/ppc/stackChunkFrameStream_ppc.inline.hpp
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/stackChunkFrameStream_ppc.inline.hpp:194]: whitespace

File: src/hotspot/cpu/ppc/templateInterpreterGenerator_ppc.cpp
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/templateInterpreterGenerator_ppc.cpp:634]: Nice cleanup!

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp
Comment (TheRealMDoerr) [inline src/hotspot/share/runtime/continuationFreezeThaw.cpp:488]: Unfortunate that we still need to distinguish, here. But, ok. I don't have a better idea atm.

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp
Comment (TheRealMDoerr) [inline src/hotspot/share/runtime/continuationFreezeThaw.cpp:1024]: Thanks for removing `unextended_sp`, here. It's not generic enough for shared code.

**Review by TheRealMDoerr — CHANGES_REQUESTED** [review]
Great work! Looks very good to me. I'm requesting precise node sizes in C2 (let's keep this invariant: nodes should either have a precisely precomputed size or use dynamic computation). The rest are just minor comments and suggestions.
I need more time to study all explanations and to review the test, but I can't see anything which prevents us from shipping it with JDK 20.
Thanks for the ASCII art pictures!

File: src/hotspot/cpu/aarch64/frame_aarch64.hpp
Comment (reinrich) [reply src/hotspot/cpu/aarch64/frame_aarch64.hpp:110]: Thanks. I don't think though that these can replace metadata_words completely because for at least one platform you would replace it with 0.

File: src/hotspot/cpu/ppc/continuationFreezeThaw_ppc.inline.hpp
Comment (reinrich) [reply src/hotspot/cpu/ppc/continuationFreezeThaw_ppc.inline.hpp]: Done

File: src/hotspot/cpu/ppc/continuationFreezeThaw_ppc.inline.hpp
Comment (reinrich) [reply src/hotspot/cpu/ppc/continuationFreezeThaw_ppc.inline.hpp]: Done

File: src/hotspot/cpu/ppc/frame_ppc.hpp
Comment (reinrich) [reply src/hotspot/cpu/ppc/frame_ppc.hpp]: Done.

File: src/hotspot/cpu/ppc/ppc.ad
Comment (reinrich) [reply src/hotspot/cpu/ppc/ppc.ad]: Done (note it's not done on other platforms).

File: src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp
Comment (reinrich) [reply src/hotspot/cpu/ppc/sharedRuntime_ppc.cpp]: Done

File: src/hotspot/cpu/ppc/stackChunkFrameStream_ppc.inline.hpp
Comment (reinrich) [reply src/hotspot/cpu/ppc/stackChunkFrameStream_ppc.inline.hpp]: Ok now?

**Review by TheRealMDoerr — APPROVED** [review]
Thanks for the updates! I'll look more into the explanations and the test, but you don't have to wait for that after you got your 2nd review.

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp
Comment (reinrich) [reply src/hotspot/share/runtime/continuationFreezeThaw.cpp:488]: A pd constant could be introduced. Not sure if it's worth it.

**Comment by reinrich** [comment]
> Thanks for the updates! I'll look more into the explanations and the test, but you don't have to wait for that after you got your 2nd review.

Thanks a lot indeed!

**Review by TheRealMDoerr — APPROVED** [review]
Thanks for mitigating the risk of test timeouts!

**Comment by reinrich** [comment]
> Thanks for mitigating the risk of test timeouts!

Yes I had to adjust the timeout a little. Also it's not necessary to set -XX:+VerifyContinuations again in the runs that call System.gc().

File: test/jdk/jdk/internal/vm/Continuation/BasicExp.java
Comment (lmesnik) [inline test/jdk/jdk/internal/vm/Continuation/BasicExp.java:729]: Please fix the indentation. There are  several places were 'i1+i2' should be fixed to 'i1 + i2 '.

**Review by lmesnik — CHANGES_REQUESTED** [review]
Not a complete review but some notes about test style.

File: test/jdk/jdk/internal/vm/Continuation/BasicExp.java
Comment (reinrich) [reply test/jdk/jdk/internal/vm/Continuation/BasicExp.java]: Ah, sure.

File: test/jdk/jdk/internal/vm/Continuation/BasicExp.java
Comment (reinrich) [reply test/jdk/jdk/internal/vm/Continuation/BasicExp.java]: > It is better to use tag @enablePreview so --enable-preview is not needed.

Done

> Also, please remove -XX:+UnlockDiagnosticVMOptions where it is not required.

The release build requires -XX:+UnlockDiagnosticVMOptions for -XX:+WhiteBoxAPI.

Without I get

```
Error: VM option 'WhiteBoxAPI' is diagnostic and must be enabled via -XX:+UnlockDiagnosticVMOptions.
Error: The unlock option must precede 'WhiteBoxAPI'.
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```

File: test/jdk/jdk/internal/vm/Continuation/BasicExp.java
Comment (reinrich) [reply test/jdk/jdk/internal/vm/Continuation/BasicExp.java:320]: Done

File: test/jdk/jdk/internal/vm/Continuation/BasicExp.java
Comment (reinrich) [reply test/jdk/jdk/internal/vm/Continuation/BasicExp.java]: Good point! Actually WB.youngGC() is sufficient and reduces the runtime significantly.

File: test/jdk/jdk/internal/vm/Continuation/BasicExp.java
Comment (reinrich) [reply test/jdk/jdk/internal/vm/Continuation/BasicExp.java]: Done

File: test/jdk/jdk/internal/vm/Continuation/BasicExp.java
Comment (reinrich) [reply test/jdk/jdk/internal/vm/Continuation/BasicExp.java:607]: Done

File: test/jdk/jdk/internal/vm/Continuation/BasicExp.java
Comment (reinrich) [reply test/jdk/jdk/internal/vm/Continuation/BasicExp.java:283]: It doesn't matter that much which compiler is used. What really matters is to have compiled frames at specific locations when freezing/thawing. The test cases control which methods are compiled and which are not. I don't want TieredCompilation to interfere with that.

**Comment by reinrich** [comment]
Thanks for your feedback @lmesnik. It help reducing the runtime of the test significantly.

File: test/jdk/jdk/internal/vm/Continuation/BasicExp.java
Comment (lmesnik) [reply test/jdk/jdk/internal/vm/Continuation/BasicExp.java]: Sorry, I meant -XX:+IgnoreUnrecognizedVMOptions where it is not needed. I think it is needed only to don't fail when -XX:+VerifyContinuations  is used in product mode and not needed in other testcases

Comment (reinrich) [reply test/jdk/jdk/internal/vm/Continuation/BasicExp.java]: Ok (I thought so). I've fixed it now.

**Comment by lmesnik** [comment]
Thanks for fixing the test.
I haven't done the full review, not adding myself as a reviewer. Don't have any additional comments.

**Comment by reinrich** [comment]
Thanks a lot for the reviews and feedback!
My own testing and GHA have succeeded after merging master.
I'm intending to integrate the port on Monday since I don't expect more feedback.

**Pull request closed without merging**
